### PR TITLE
19494 Remove no-arg constructor requirement for no-configuration use case

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Fixes 19494 Remove no-arg constructor requirement for no-configuration use case by using Objenesis library.

Checklist:
- [ ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [ ] Request reviewers if possible
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
Thanks, I've fixed the checkstyle and remove that fail-fast method

**Comment by mdumandag** [comment]
@frant-hartm, it looks all good to me, but when you have the time, can you take a look at the shading of the newly added dependency and comment if there are any problems with that? Just to make sure there are no problems with it. Thanks

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java [hunk 1]
```diff
@@ -259,7 +260,7 @@ private CompactSerializableRegistration getOrCreateRegistration(String typeName)
                 return null;
             }
             try {
-                Object object = ClassLoaderUtil.newInstance(clazz.getClassLoader(), clazz);
+                Object object = newInstanceBypassingConstructor(clazz);
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java:263]: If I understand this correctly then it skips the constructor in all cases, not only if it's not available, correct? Are we sure we want to do it this way? Maybe we can fallback to bypassing constructor only if the default constructor doesn't exist?

**Review by frant-hartm — APPROVED** [review]
The shading part looks all right, I have checked also the source jar and it looks fine.

One unrelated question, but not a blocker.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java (hunk 1 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java:263]: Hi, yes, it skips constructor in all cases. I thought it's the easiest way, but can change it too - no preferences here from my side

File: hazelcast/src/main/java/com/hazelcast/internal/nio/InstanceCreationUtil.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.function.Supplier;
+
+import static com.hazelcast.internal.util.ExceptionUtil.rethrow;
+
+/**
+ * Utility class to deal with creation of new object instances.
+ */
+public final class InstanceCreationUtil {
+
+    private static final Objenesis OBJENESIS = new ObjenesisStd();
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/nio/InstanceCreationUtil.java:34]: Why did you not use [this](http://objenesis.org/apidocs/org/objenesis/ObjenesisStd.html#ObjenesisStd-boolean-) constructor which does do the caching for the underlying instantiator (no need for `OBJECT_SUPPLIERS` in that case)?

**Comment by TomaszGaweda** [comment]
Hi @sancar, changes are ready

**Comment by sancar** [comment]
run-lab-run

**Review by sancar — APPROVED** [review]

**Review by ihsandemir — APPROVED** [review]

**Comment by sancar** [comment]
@TomaszGaweda Thanks for the pr. I am merging it.

**Pull request merged**